Caching & Fork Support

criticic opened the pull request.

criticic commented:
for usage would require a few additional changes in the packages that use it

https://github.com/criticic/data.table-test-work-workflow/blob/master/.github/workflows/performance-tests.yml

- pull_request -> pull_request_target
- passing github-token to the action

Anirban166 commented on action.yml:
```diff
@@ -25,33 +30,103 @@ runs:
+        git switch -c "${{ github.head_ref }}" "fork/${{ github.head_ref }}"
+        # Ensure base ref is also available locally
+        git switch "${{ github.base_ref }}"
+        git switch "${{ github.head_ref }}"
```
I can see that a multi-step workaround is necessary here given the runner only has a detached HEAD (pointing to a commit in the PR / not tracking the fork's branch) but this looks somewhat convoluted (assuming this is perfectly aligned with the goal of this step - for `git2r` to find the local branch references for forked branches as well)

At the very least, I would consider it a candidate for simplification given the plethora of git commands being used. For e.g., instead of adding a remote for the fork, fetching it, and then creating a named branch locally, you can directly fetch the PR's head into a local branch using:
```sh
git fetch origin pull/${{ github.event.pull_request.number }}/head:${{ github.head_ref }}
```

Anirban166 commented on action.yml:
```diff
@@ -25,33 +30,103 @@ runs:
+      id: download-r-lib
+      uses: dawidd6/action-download-artifact@v10 
+      with:
+        name: r-lib-cache-${{ hashFiles('**/DESCRIPTION') }}
```
This file-dependent artifact naming can potentially lead to bloating. I'm guessing this since the name is being generated based on the hash of the target repository's `DESCRIPTION` file, meaning an artifact will be orphaned with the old hash (and thus remain in that repo's storage) every time changes are made to the `DESCRIPTION` (even minor ones like updating the version number say), unless the artifacts are tied to some retention rules (e.g. automatic deletion of cached artifacts that are left unused for x amount of time)

Anirban166 commented on action.yml:
```diff
@@ -25,33 +30,103 @@ runs:
+          atime
+          ggplot2
+          directlabels
+          cumstats=?ignore
```
Is this an accidental inclusion? 
I don't remember this package being required, but if it is indeed a necessity, why is it configured to be ignored?

criticic replied:
cumstats(a suggested package of atime) was removed from CRAN, earlier we installed these three packages using install.packages, but setup-r-dependencies fails whenever it cannot find any optional dependency, so had to ignore it explicitly

criticic replied:
Yeah, I just wanted to install all packages in single step with pak (also was hoping to use the pkg.lock generated by pak for the cache hash), i can revert this change

Anirban166 replied:
If using `pak`/`pkg.lock` leads to time savings then I would recommend continuing, else revert.

In https://github.com/r-lib/actions/tree/v2/setup-r-dependencies#usage I see the default for `dependencies` is set to `"all"`, so have you tried this:
```yaml
    - name: Install R packages (DESCRIPTION-specified deps)
      if: steps.cache-r-lib.outputs.cache-hit != 'true'
      uses: r-lib/actions/setup-r-dependencies@v2
      with:
        cache: false
        dependencies: "hard"
        extra-packages: |
          atime
          ggplot2
          directlabels
```

Anirban166 commented on action.yml:
```diff
@@ -25,33 +30,103 @@ runs:
-        echo "options(repos = c(CRAN = 'https://cloud.r-project.org'))" >> .Rprofile
-        Rscript -e 'install.packages(c("atime", "ggplot2", "directlabels"))'
+        mkdir -p ~/.R
+        echo "CFLAGS = -std=gnu99" >> ~/.R/Makevars
```
Likewise, why is this C compiler version globally set? (or C99-compliant code specifically enforced?)

Anirban166 commented on action.yml:
```diff
@@ -25,33 +30,103 @@ runs:
+        mkdir -p /tmp
+        sudo tar -czf /tmp/libgit2-cache.tar.gz \
+          /usr/local/lib/libgit2* \
+          /usr/local/include/git2* 2>/dev/null || true
```
Redirecting to `/dev/null` and forcing a `true` exit code (with the logical or) is a shortcut to masking errors and losing visibility when something goes wrong in this step and further along the line subsequently.

Anirban166 commented on action.yml:
```diff
@@ -84,16 +179,35 @@ runs:
+
+        R_SETUP_FORMATTED=$(formatTime $R_SETUP_DURATION)
+        VERSION_SETUP_FORMATTED=$(formatTime $VERSION_SETUP_DURATION)
+        ATIME_TESTS_FORMATTED=$(formatTime $ATIME_TESTS_DURATION)
```
I wouldn't create separate variables for just holding the formatted version (also doubles the number, 6 instead of 3 here), and having `DURATION` omitted from their name makes it less clear to identify what these mean or store at a glance (name-wise, `R_SETUP_DURATION` > `R_SETUP_DURATION_FORMATTED` > `R_SETUP_FORMATTED`)

Anirban166 commented on action.yml:
```diff
@@ -84,16 +179,35 @@ runs:
+        ATIME_TESTS_FORMATTED=$(formatTime $ATIME_TESTS_DURATION)
+
+        # Start building the report markdown
+        REPORT="## atime Performance Analysis Results\n\n"
```
I like the comment without a heading since it doesn't add any extra information, also looks a bit better to me that way

Anirban166 commented on action.yml:
```diff
@@ -84,16 +179,35 @@ runs:
+        REPORT="${REPORT}|---------------------------------------|-------------------------|\n"
+        REPORT="${REPORT}| R setup and installing dependencies   | ${R_SETUP_FORMATTED}    |\n"
+        REPORT="${REPORT}| Installing different package versions | ${VERSION_SETUP_FORMATTED} |\n"
+        REPORT="${REPORT}| Running and plotting the test cases   | ${ATIME_TESTS_FORMATTED}  |"
```
Not sure why you're building a `REPORT` string by chaining but starting every line here with `REPORT="${REPORT}...` looks less clean than the `echo` statements I had there previously. Now that I'm thinking, a better approach would be to use a heredoc or `cat << EOF ... EOF`.

Why are you making changes to the format of the comment anyway? 
It's outside the scope of this PR (If you wanted to discuss changes for that, it should be in a separate PR)

Anirban166 commented on action.yml:
```diff
@@ -84,16 +179,35 @@ runs:
+        REPORT="${REPORT}|---------------------------------------|-------------------------|\n"
+        REPORT="${REPORT}| R setup and installing dependencies   | ${R_SETUP_FORMATTED}    |\n"
+        REPORT="${REPORT}| Installing different package versions | ${VERSION_SETUP_FORMATTED} |\n"
+        REPORT="${REPORT}| Running and plotting the test cases   | ${ATIME_TESTS_FORMATTED}  |"
```
Not sure why you're building a `REPORT` string by chaining but starting every line here with `REPORT="${REPORT}...` looks less clean than the `echo` statements I had there previously. Now that I'm thinking, a better approach would be to use a heredoc or `cat << EOF ... EOF`

Why are you making changes to the format of the comment anyway? 
It's outside the scope of this PR (If you wanted to discuss changes for that, it should be in a separate PR)

Anirban166 replied:
Also, the amount of comments you have here should have been in your code above (for steps such as retrieving branch references and installing `libgit2`), the comments in this part (related to GH-bot comment contents) are sorta unnecessary given the comparatively obvious nature of things

Anirban166 reviewed (commented):
Needs changes and explanations, but good to see some progress (more specifically to see the caching being effective) 

(Also, ping @tdhock to weigh in)

tdhock commented:
why caching and fork support in the same PR?
typically I would suggest two different PRs for two different features.

criticic commented on action.yml:
```diff
@@ -9,6 +9,19 @@ inputs:
+    description: 'The specific commit SHA of the head ref'
+    required: true
+  pr-number:
+    description: 'The pull request number, if the trigger is a PR. Leave empty for other triggers.'
```
I was planning on allowing different triggers as well, so as to have it run on the base branch

Anirban166 replied:
The message seems reasonable then, but keep in mind this action is specifically meant to be triggered on PRs only, so although I recognize the need to go this route or use `workflow_dispatch` to help with the initial setup in establishing a cache for the target repo's base branch (e.g. data.table's master), I don't think we would be using other triggers after the seeding run (where we expect a cache miss as a one-time thing, but forthcoming runs would fetch the saved libraries/packages, and for all of those subsequent triggers I would expect them to be PRs to that same branch)

criticic commented:
Ran via PR (cached): https://github.com/criticic/data.table-test-work-workflow/actions/runs/15947154457 (took around 4m)

Ran Manually using Github Web UI (`workflow_dispatch`): https://github.com/criticic/data.table-test-work-workflow/actions/runs/15946446968/job/44981117330 (libgit was already cached in this run, so 10ms rather than around 13m)

The Caches Saved scoped to `master`: https://github.com/criticic/data.table-test-work-workflow/actions/caches

this is how we call this action from the workflow: https://github.com/criticic/data.table-test-work-workflow/blob/master/.github/workflows/performance-tests.yml

Anirban166 commented on action.yml:
```diff
@@ -64,7 +134,8 @@ runs:
         path: .ci/atime/
 
-    - name: Push generated plot along with relevant information on the PR thread as a GH-bot comment
+    - name: Generate report markdown
```
Generate report

Anirban166 commented on action.yml:
```diff
@@ -5,6 +5,24 @@ branding:
+  pr-number:
+    description: 'The pull request number, if the trigger is a PR. Leave empty for other triggers.'
+    required: false
+    default: ''
```
Since this is already specified as empty by default here, you can likely remove [this line](https://github.com/criticic/data.table-test-work-workflow/blob/3104b08bacba7272b93dc7897403bb4bcec603fe/.github/workflows/performance-tests.yml#L44) (seems redundant) as when an output variable is not set by a step, its value effectively becomes an empty string or null in the GitHub Actions expression context iirc

Anirban166 commented on action.yml:
```diff
@@ -88,12 +159,29 @@ runs:
+      run: |
+        echo "## Atime Report (Manual Dispatch) ##"
+        cat report.md
+      shell: bash
```
Also, remove this last step as it won't get much use (only the first run realistically, and I doubt a report in the console would be appropriate; nice to have maybe but not necessary)

Anirban166 reviewed (changes requested):
Minor changes needed for my end of this review and I wish the calling workflow was less verbose, but LGTM overall! 

Going ahead though, don't forget to show/link an example GHA run when updating code (especially with substantial changes) for quick reference, and avoid force-pushing.

@tdhock what do you think of these changes?

tdhock commented:
"logic about old vs new compiler version/flags" code should probably go in atime::atime_versions (not only in GHA)

criticic committed:
feat: Add Caching & Support Non-PR workflow events

criticic commented:
Some Test Runs (no major changes since last time, but now the downstream action is cleaner)

Without Caching (Run using the Github Web UI): https://github.com/criticic/data.table-test-work-workflow/actions/runs/16079721267 took 10m 58s

Cached Files: https://github.com/criticic/data.table-test-work-workflow/actions/caches takes about 78MB of storage

With Caching (PR from non-fork): https://github.com/criticic/data.table-test-work-workflow/actions/runs/16079890711 took 3m 14s

With Caching (PR from fork): https://github.com/criticic/data.table-test-work-workflow/actions/runs/16079900324 took 3m 37s

tdhock commented:
this sounds reasonable, thanks for the explanation.

tdhock commented:
can you please push some changes to the README or other docs which explain what new users have to do?

criticic committed:
Update Readme

tdhock commented:
please add documentation about how to manually invalidate the entire cache

criticic commented:
The would just involve deleting the cache from the web ui https://github.com/criticic/data.table-test-work-workflow/actions/caches or using the gh cli

Should I add this information?

On 11 Jul 2025, at 1:11 AM, Toby Dylan Hocking ***@***.***> wrote:

[https://avatars.githubusercontent.com/u/932850?s=20&v=4]tdhock left a comment (Anirban166/Autocomment-atime-results#44)<https://github.com/Anirban166/Autocomment-atime-results/pull/44#issuecomment-3058779757>

please add documentation about how to manually invalidate the entire cache

—
Reply to this email directly, view it on GitHub<https://github.com/Anirban166/Autocomment-atime-results/pull/44#issuecomment-3058779757>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AJVJDCVSHAYH57E6VITMIID3H26VTAVCNFSM6AAAAACAAYF6BWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTANJYG43TSNZVG4>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

tdhock commented:
yes please add info for both options

criticic committed:
docs: Add cache invalidation and manual clearing instructions to Readme

Anirban166 commented on Readme.md:
```diff
@@ -1,5 +1,7 @@
 A GitHub Action for R packages that runs `atime::atime_pkg` and comments the generated results on pull requests to help identify potential performance regressions introduced from the incoming changes.
 
+This action now includes **caching** for R packages and system dependencies to significantly speed up subsequent workflow runs. Due to this, there are significant changes in how the action is used, as detailed [below](#caching).
```
I'd avoid double asterisks or the need to emphasize a word unless it carries critical meaning for the user to take note of (plus you already have a section for caching below)

Two more suggestions for this para:
- 'incorporates' instead of 'includes' fits better imo
- Avoiding the use of same words is something I try to do, so maybe use a different word for one of the two 'significant' terms (for e.g. 'Consequently, there are notable changes ... ' over 'Due to this, there are significant changes ... ')

Suggestion:
'This action now incorporates caching for R packages and system dependencies to significantly speed up subsequent workflow runs. Consequently, there are notable changes in how the action is used, as detailed [below](https://www.google.com/search?q=%23caching).

Anirban166 commented on Readme.md:
````diff
@@ -33,14 +35,67 @@ on:
+        with:
+          REPO_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+```
+Emplace the contents in `.github/workflows/<workflowName>.yml`. The example above can be customized, but a few things are important for it to work correctly:
````
'... but a few things are important <insert?> for it to work ...'
important to what? take note of?

What I initially had here was 'The example I provided above can be customized further as needed, as long as a few things are kept intact:' where I'm telling the user to retain some things as is for a similar <insert>, so please make sure your use of language is correct when changing writing.

Anirban166 replied:
'... but a few things are important <insert?> for it to work ...'
important to what? take note of?

What I initially had here was 'The example I provided above can be customized further as needed, as long as a few things are kept intact:', so please make sure your use of language when changing is correct or considers retaining the same clarity and flow.

Anirban166 commented on Readme.md:
````diff
@@ -33,14 +35,67 @@ on:
+          REPO_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+```
+Emplace the contents in `.github/workflows/<workflowName>.yml`. The example above can be customized, but a few things are important for it to work correctly:
+- The workflow should run on a **`pull_request_target`** event. This allows the action to run in the context of the base repository, which is necessary for it to have permission to post comments securely.
````
Same here and below, avoid the bold emphasis unless necessary

Anirban166 reviewed (commented):
Thanks! (also great suggestion by Toby as imo updating user-oriented documentation for every feature or code change is essential, especially major ones like this)

criticic committed:
Fixes

criticic committed:
use only dependencies + tests.R as hash

criticic committed:
also update the readme

criticic committed:
change to 3rd person pov

criticic commented on Readme.md:
```diff
@@ -33,14 +35,67 @@ on:
+
+- You modify the `DESCRIPTION` file of your R package.
+
+Also note [Github's Cache Usage limits and eviction policy](https://docs.github.com/en/actions/reference/dependency-caching-reference#usage-limits-and-eviction-policy), which states that GitHub will remove any cache entries that have not been accessed in over 7 days. There is no limit on the number of caches you can store, but the total size of all caches in a repository is limited to 10 GB. Once ...
```
sorry mised this one

criticic committed:
fix

tdhock commented on Readme.md:
````diff
@@ -33,14 +35,68 @@ on:
+  pull_request_target:
+    types: [opened, reopened, synchronize]
+```
+When a commit is pushed to `main`, the workflow will run, install all dependencies, and save them to the cache. Subsequent PRs will then be able to download this cache, skipping the slow installation steps.
````
"install all dependencies, and save them to the cache" is not clear.
A dependency would be something that is required to run this action, for example atime. Is that cached?
I thought the main point was to cache the historical versions of the target R package that will be tested. Can this point be clarified please? (please explain what exactly will be cached)

criticic replied:
```md
When the workflow runs on the base branch for the first time, it will install all dependencies and save them to the cache. Specifically, the cache includes:

- The entire R library path (`.libPaths()[1]`), which contains:
  - Historical versions of the target R package being tested
  - Dependencies such as `atime`, `ggplot2`, and `directlabels`
- The built `libgit2` library, which is a dependency of the `git2r` package, required by `atime`

Subsequent PRs will then be able to download this cache, skipping the slow installation steps for both the dependencies and the historical versions of the target package.
```

is this better?

criticic committed:
clarify the files saved

tdhock commented:
this looks good to me now. please merge when ready.
